worker/src/RTC/RtpStreamSend.cpp

```diff
@@ -36,107 +37,162 @@ namespace RTC
-		auto addToBack{ static_cast<uint16_t>((seq - (this->startSeq + this->buffer.size() - 1))) %
-			              MaxSeq };
-		// Calculate how many elements would it be necessary to add when pushing new item to the front
-		// of the deque.
-		auto addToFront{ static_cast<uint16_t>((this->startSeq - seq) % MaxSeq) };
+		auto bucket = this->buckets[bucketIdx];
 
-		// Select the side of deque where fewer elements need to be added, while preferring the end.
-		if (addToBack <= addToFront)
+		if (!bucket)
 		{
-			// Packets can arrive out of order, add blank slots.
-			for (uint16_t i{ 1 }; i < addToBack; ++i)
-				this->buffer.push_back(nullptr);
+			return false;
+		}
 
-			this->buffer.push_back(storageItem);
+		if (!bucket->Get(inBucketPos))
+		{
+			return false;
 		}
-		else
+
+		auto item = bucket->Remove(inBucketPos);
+		resetStorageItem(item);
+		StorageItemPool.Return(item);
+
+		if (bucket->IsEmpty())
 		{
-			// Packets can arrive out of order, add blank slots.
-			for (uint16_t i{ 1 }; i < addToFront; ++i)
-				this->buffer.push_front(nullptr);
+			StorageItemBucketPool.Return(bucket);
+			this->buckets[bucketIdx] = nullptr;
+		}
 
-			this->buffer.push_front(storageItem);
-			this->startSeq = seq;
+		if (this->oldestSeq == seq)
+		{
+			this->InvalidateOldest(seq);
 		}
 
 		return true;
 	}
 
-	bool RtpStreamSend::StorageItemBuffer::Remove(uint16_t seq)
+	void RtpStreamSend::StorageItemBuffer::Clear()
 	{
-		if (this->buffer.empty())
-			return false;
+		for (auto& bucket : this->buckets)
+		{
+			if (!bucket)
+				continue;
 
-		auto idx{ static_cast<uint16_t>((seq - this->startSeq) % MaxSeq) };
+			for(size_t pos = 0; pos < bucket->Size() && !bucket->IsEmpty(); ++pos)
+			{
+				auto item = bucket->Remove(pos);
 
-		this->buffer[idx] = nullptr;
+				if (item)
+				{
+					resetStorageItem(item);
+					StorageItemPool.Return(item);
+				}
+			}
 
-		// If we have erased the first element, remove all `nullptr` elements from the beginning of the buffer.
-		if (idx == 0)
+			StorageItemBucketPool.Return(bucket);
+			bucket = nullptr;
+		}
+
+		this->oldestSeq = -1;
+	}
+
+	RtpStreamSend::StorageItem* RtpStreamSend::StorageItemBuffer::GetOldest() const
+	{
+		if (this->oldestSeq >= 0)
 		{
-			while (!this->buffer.front())
-			{
-				this->buffer.pop_front();
-				this->startSeq++;
-			}
+			return this->Get(static_cast<uint16_t>(this->oldestSeq));
 		}
-		// If we have erased the last element, remove all `nullptr` elements from the end of the buffer.
-		else if (idx == static_cast<uint16_t>(this->buffer.size() - 1))
+
+		return nullptr;
+	}
+
+	void RtpStreamSend::StorageItemBuffer::TryUpdateOldest(uint16_t seq, RtpStreamSend::StorageItem* storageItem)
+	{
+		auto oldest = this->GetOldest();
+
+		if (!oldest)
 		{
-			while (!this->buffer.back())
-				this->buffer.pop_back();
+			this->oldestSeq = seq;
+			return;
 		}
 
-		return true;
+		if (oldest->originalPacket->GetTimestamp() > storageItem->originalPacket->GetTimestamp())
```

Comment:
We cannot compare timestamps this way since the value can be wrapped around. Use this:

```c++
RTC::SeqManager<uint32_t>::IsSeqHigherThan(oldest->originalPacket->GetTimestamp(), storageItem->originalPacket->GetTimestamp())
```

Reply:
done